hera_cal/abscal.py

```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+    else:
+        # if no autocorrelation file supplied, use nsamples weights times
+        # or of data and model flags
+        wgts = DataContainer({k: (~data_flags[k]).astype(np.float64)
+                             * (~model_flags[k]).astype(np.float64)
+                             * data_nsamples[k] for k in data_flags})
+
+    # select refant if not specified.
+    if refant is None:
+        refant = pick_reference_antenna(abscal_gains, abscal_flags, hc.freqs, per_pol=True)
+    hc.ref_antenna_name = str(refant)
+    hcm.ref_antenna_name = str(refant)
+
+    abscal = AbsCal(data=data, model=model, wgts=wgts, freqs=data.freqs)
+
+    delta = 9e99
+
+    abscal_gains_iteration = {}
+
+    # find initial starting point with lincal before performing abscal.
+    if dly_lincal:
+        abscal.delay_lincal()
+        for k in abscal_gains:
+            abscal_gains_iteration[k] = abscal.ant_dly_phi_gain[k] * abscal.ant_dly_gain[k]
+        apply_cal.calibrate_in_place(data=abscal.data, new_gains=abscal_gains_iteration)
+        for k in abscal_gains:
+            abscal_gains[k] *= abscal_gains_iteration[k]
+
+    niter = 0
+    while delta > tol and niter < max_iter:
+        if not no_ampcal:
+            abscal.amp_logcal(verbose=verbose)
+        if not no_phscal:
+            abscal.phs_logcal(verbose=verbose)
+        for k in abscal.ant_eta_gain:
+            abscal_gains_iteration[k] = np.ones_like(abscal_gains[k])
+            if not no_ampcal:
+                abscal_gains_iteration[k] *= abscal.ant_eta_gain[k]
+            if not no_phscal:
+                abscal_gains_iteration[k] *= abscal.ant_phi_gain[k]
+        # phase to refant.
+        rephase_to_refant(abscal_gains_iteration, refant, flags=abscal_flags, propagate_refant_flags=True)
+        # update abscal gains with iteration.
+        abscal_gains_new = {k: abscal_gains[k] * abscal_gains_iteration[k] for k in abscal_gains}
+        delta = np.max([np.max(np.abs(abscal_gains_new[k][np.invert(abscal_flags[k])]
+                        - abscal_gains[k][np.invert(abscal_flags[k])])) for k in abscal_gains])
+
+        for k in abscal_gains:
+            abscal_gains[k] = abscal_gains_new[k]
+
+        # use hcm to apply iteration gains.
+        hcm.update(gains=abscal_gains_iteration)
+
+        # apply iteration gains to abscal.data
+        apply_cal.calibrate_in_place(data=abscal.data, new_gains=abscal_gains_iteration)
+        niter += 1
+        if delta <= tol:
+            echo(f"Convergence achieved after {niter} iterations with max delta of {delta}", verbose=verbose)
+        if niter == max_iter - 1 and delta > tol:
+            echo(f"Convergence not achieved but max_iter of {max_iter} reached. delta={delta}.", verbose=verbose)
+
+    # update the calibration array.
+    hc.update(gains=abscal_gains)
+
+    # multiply gains by precal gains
+    if precalibration_gain_file is not None:
+        hc.gain_array *= uvc_precal.gain_array
```

Comment:
I think it may be possible that the gain_array antenna ordering in `hc` is different than that of `uvc_precal`. I would suggest eliminating this possible failure mode by either
1. Iterating through hc.ant_array and updating each antenna/pol individually, or 
2. (easier) using `merge_gains([])` to update `abscal_gains` dictionary with `uvc_precal.gain_array` in dictionary form before using `hc.update(gains=abscal_gains)`

Reply:
I've changed this to use `merge_gains`